test/observer/toast.dart

```diff
@@ -0,0 +1,32 @@
+import 'package:flutter/services.dart';
+
+import '../test.dart';
+
+/// An observer for toast messages from the flutter toast package.
+class ToastChannelObserver {
+  /// The method channel used by the flutter toast package
+  static const MethodChannel _channel = MethodChannel('PonnamKarthik/fluttertoast');
+  Map<String, dynamic>? _lastToastArguments;  /// The arguments for the last requested toast message.
+  String? get lastToastMessage {
+    final lastArguments = _lastToastArguments;
+    clearLastToast();
+    return lastArguments?['msg'] as String?;
+  }
+  
+  /// Create a new toast channel observer, which automatically
+  ///  unregisters any previously created observer.
+  ToastChannelObserver(WidgetTester tester) {
+    tester.binding.defaultBinaryMessenger.setMockMethodCallHandler(_channel, (call) {
+      if (call.method == 'showToast') {
+        _lastToastArguments = Map.castFrom(call.arguments);
+        return null;
+      }
+      return null;  // Ignore unimplemented method calls
+    });
+  }
+
+  /// Forget the last received toast.
+  void clearLastToast() {
+    _lastToastArguments = null;
```

Comment:
This can be removed

Reply:
Being able to clear the last toast message is very important in my opinion. Consider this:

``` dart
... // Code that should generates Toast 1.
expect(toastObserver.lastToastMessage, l10n.toast1);
... // Code that should generate Toast 2.
expect(toastObserver.lastToastMessage, l10n.toast2);
```

If we don't clear the `lastToastMessage` after checking for `toast1`, it leaks in the second part of the test. If the code that should generate `toast2` doesn't generate a toast, we wan't the second expect to fail and tell us that no toast was displayed, which is unequal to the expected `toast2`.
Instead, we get `toast1` is unequal to `toast2`, which is confusing and doesn't point us in the direction of what's going wrong here. Even worse, consider if `toast1` and `toast2` were actually the same string, for example if I want to test that the exit confirmation toast shows again when I wait 2 or more seconds. In that case, even if the second code doesn't generate the toast, the except still passes, because `lastToastMessage` contains the `toast1` from the first code. So the test becomes useless and even harmful at this point, because we think we are testing something that we are actually not.

I strongly believe that we must clear in order to protect us from situations like this. Now, one might argue that we could solve this her by just keeping a list of all recorded toast messages. But this only works for this case because we are handling strings, and most importantly, this makes the test more complicated and dependent on what was done before. If we decides to add an additional part before this test, the history of toasts changes. But in our test, we are only ever interested about the history since we last observed last, or since we last cleared.

Reply:
I believe the approach you have chosen with keeping only last message is error-prone and less clear.

> But this only works for this case because we are handling strings, and most importantly, this makes the test more complicated and dependent on what was done before. 

This would work with any matcher, not only strings, it doesn't make the test any more complicated. 

> If we decides to add an additional part before this test, the history of toasts changes. But in our test, we are only ever interested about the history since we last observed last, or since we last cleared.

That would be perfect if we notice that our test is affected by something we didn't notice on our own and change the test accordingly to expect it, or investigate why this is and fix what affected it.

Reply:
Also this kind of magic of clearing the state after calling a getter is what should be avoided in any kind of code

Reply:
> I believe the approach you have chosen with keeping only last message is error-prone and less clear.

I disagree, the current approach avoids potential errors like the one I showed above. If we keep the history, I need an additional check to detect the fact that there was no toast generated:
```dart
... // Code that should generates Toast 1.
expect(toastObserver.toastMessages.last, l10n.toast1);
var numToastsBefore = toastObserver.toastMessage.length;
... // Code that should generate Toast 2.
expect(toastObserver.toastMessage.length, numToastsBefore + 1);
expect(toastObserver.lastToastMessage, l10n.toast2);
```

This not only makes the test more complicate to read, it also generates an additional burden for us when we write the tests, because we need to remember to add this check, otherwise our tests will not test what we actually think they test.

Reply:
> That would be perfect if we notice that our test is affected by something we didn't notice on our own and change the test accordingly to expect it, or investigate why this is and fix what affected it.

The problem is that if we do that, we are not testing that the app behaves as expected, we are testing that out test behaves exactly as we originally thought it would. The failing tests don't indicate a problem in the app, just a change in the test. I don't see the benefit in that, because it makes our tests less flexible, and it moves our focus away from the app to the test. It could also cause us to think that a failing test is caused by a change we do to the test, when in reality it uncovers a bug in the actual app, but we then "fix" the test, because we think the failure is caused by our change in the test before.

Reply:
> Also this kind of magic of clearing the state after calling a getter is what should be avoided in any kind of code

I do agree with that, how about this:

I remove the `lastToastMessage` getter and instead add a function `popLastToastMessage`. Then the intent would be clear.

Reply:
> The problem is that if we do that, we are not testing that the app behaves as expected, we are testing that out test behaves exactly as we originally thought it would.

No, we are cementing in a test that our app shown exactly those toasts that we expected

```dart
expect(toastObserver.toastMessagesLog, []);

functionThatShowsOneToast();
expect(toastObserver.toastMessagesLog, [l10n.toast1]);

someFunctionThatShouldOnlyShowOneMoreAdditionalToast();
expect(toastObserver.toastMessagesLog, [l10n.toast1, l10n.toast2]);
```

If at some point `someFunctionThatShouldOnlyShowOneMoreAdditionalToast` starts to show N additional toasts that we didn't expect, with your approach we fail to notice that. Checking it like above will make the test fail and we will investage why is that - if expected, change the test accordingly, otherwise fix `someFunctionThatShouldOnlyShowOneMoreAdditionalToast`.

I insist that it's done like this.

Reply:
Ok, I see now that your approach is definitely the only way to deal with that approach, and it does make the tests more explicit, which they should be. I made the changes in the latest commit.